daemon/containerd/image_list.go

```diff
@@ -91,12 +91,26 @@ func (i *ImageService) Images(ctx context.Context, opts types.ImageListOptions)
 			return nil, err
 		}
 
+		repoTag := "<none>:<none>"
+		repoDigest := "<none>@<none>"
+		target := img.Target().Digest
+
+		ref, err := reference.ParseNamed(img.Name())
+		if err == nil {
```

Comment:
- Do we know in what case this could produce an error? (i.e., how we can end up with an image with an invalid name?)
- IIUC, previously, we would show `img.Name()` (even if that could theoretically be an invalid name); in the new situation we replace that with `<none>:<none>` (discarding the name altogether). I'm curious in what situations that would happen (as it means we're loosing  information about the image in the response).

Reply:
IMO, there are 2 scenarios:
1) User fiddles with the `moby` containerd namespace directly and creates an image with name that's not a valid reference.
2) We make some bug/error which creates image with bad name.

In both cases, I don't think we can guarantee that referencing these images by their name will succeed (as we do `reference.ParseNormalizedName` all over the place), so not presenting these names seems reasonable to me.

Reply:
> User fiddles with the moby containerd namespace directly and creates an image with name that's not a valid reference.
 
- First of all, not sure if that's a scenario we must "ignore". Does containerd accept image names that we do not accept valid? (e.g. can I create an image in containerd that's named `🐳😄🐱` ?)
- Second: if this happens, the current code would silently ignore this case
- And (finally): if this happens: how can I remove said image? (if none of our code is able to handle the invalid name, then A) `docker image rm` wouldn't be able to delete it?, and B) this output does not allow me to "discover" the invalid name (because we hide it), which does not allow the user to correct their "fiddling" by removing it from `containerd` (as they wouldn't know _what_ `<none>:<none>` image this is).

I think currently it could even mean that an image ends up with _multiple_ `<none>:<none>` tags (one for each invalid image name). Which, thinking of that scenario, could be problematic on other ways (it's not dangling, as we only consider dangling if there's a _single_ `<none>:<none>` reference)

> We make some bug/error which creates image with bad name.

This definitely sounds like something we should not silently ignore.

Reply:
> * First of all, not sure if that's a scenario we must "ignore". Does containerd accept image names that we do not accept valid? (e.g. can I create an image in containerd that's named `🐳😄🐱` ?)

Yeah, it happily does.

```bash
$ ctr --address /run/docker/containerd/containerd.sock -n moby image ls
REF                                                                              TYPE                                                      DIGEST                                                                  SIZE    PLATFORMS                                                                                LABELS
docker.io/library/alpine:latest                                                  application/vnd.docker.distribution.manifest.list.v2+json sha256:f271e74b17ced29b915d351685fd4644785c6d1559dd1f2d4189a5e851ef753a 3.1 MiB linux/386,linux/amd64,linux/arm/v6,linux/arm/v7,linux/arm64/v8,linux/ppc64le,linux/s390x -
$ ctr --address /run/docker/containerd/containerd.sock -n moby image tag docker.io/library/alpine:latest 🐳😄🐱
🐳😄🐱
$ ctr --address /run/docker/containerd/containerd.sock -n moby image ls
REF                                                                              TYPE                                                      DIGEST                                                                  SIZE    PLATFORMS                                                                                LABELS
docker.io/library/alpine:latest                                                  application/vnd.docker.distribution.manifest.list.v2+json sha256:f271e74b17ced29b915d351685fd4644785c6d1559dd1f2d4189a5e851ef753a 3.1 MiB linux/386,linux/amd64,linux/arm/v6,linux/arm/v7,linux/arm64/v8,linux/ppc64le,linux/s390x -
🐳😄🐱                                                                              application/vnd.docker.distribution.manifest.list.v2+json sha256:f271e74b17ced29b915d351685fd4644785c6d1559dd1f2d4189a5e851ef753a 3.1 MiB linux/386,linux/amd64,linux/arm/v6,linux/arm/v7,linux/arm64/v8,linux/ppc64le,linux/s390x -
$ docker image ls
REPOSITORY   TAG       IMAGE ID       CREATED          SIZE
alpine       latest    f271e74b17ce   44 seconds ago   3.26MB
<none>       <none>    f271e74b17ce   3 seconds ago    3.26MB

```

The image will still be shown, and can be referenced by the ID. In this case the ID will be ambiguous because we tagged it from alpine, so he won't be able to delete this 🐳😄🐱 image by using Docker.

We could use the `isDanglingImage` from https://github.com/moby/moby/pull/44781 here and only override the repo tags to `<none>:<none>` for our special "dangling" images and still persist the non-standard names.

> This definitely sounds like something we should not silently ignore.
Good point; warning would be useful here

Reply:
Bumping this; I think this PR is ready to bring in once these concerns are addressed.

Reply:
I added some error logging and removed overriding the tag with `<none>` for non-dangling images.
```
ERRO[2023-01-30T14:20:45.364773471Z] failed to parse image name as reference       digest="sha256:f271e74b17ced29b915d351685fd4644785c6d1559dd1f2d4189a5e851ef753a" error="invalid reference format" name="🐳😄🐱"
```

Reply:
As for images with invalid name:

> IIUC, previously, we would show img.Name() (even if that could theoretically be an invalid name); in the new situation we replace that with <none>:<none> (discarding the name altogether). I'm curious in what situations that would happen (as it means we're loosing information about the image in the response).

Previous code did set the `repoTags={img.Name()}` but doing that made the CLI not show them **at all**. This is because CLI attempts to parse the `repoTags` as a named reference (it needs to be able to show the `REPO` and `TAG`), and ignores the image if it fails to.
https://github.com/docker/cli/blob/4a500f690f2ef80707839d7b0bbbf38b46b36991/cli/command/formatter/image.go#L116-L119

So, keeping the previous approach (repoTags=invalid name) won't show them in the `docker images` output at all:

```bash
$ ctr --address /run/docker/containerd/containerd.sock -n moby image tag docker.io/library/alpine:l🐳😄🐱🐳😄🐱
🐳😄🐱

$ docker image ls
REPOSITORY   TAG       IMAGE ID       CREATED          SIZE
alpine       latest    f271e74b17ce   23 seconds ago   3.26MB

$ curl -s --unix-socket /var/run/docker.sock http://localhost/images/json | jq
[
  {
    "Containers": -1,
    "Created": 1675087679,
    "Id": "sha256:f271e74b17ced29b915d351685fd4644785c6d1559dd1f2d4189a5e851ef753a",
    "Labels": null,
    "ParentId": "",
    "RepoDigests": [
      "docker.io/library/alpine@sha256:f271e74b17ced29b915d351685fd4644785c6d1559dd1f2d4189a5e851ef753a"
    ],
    "RepoTags": [
      "docker.io/library/alpine:latest"
    ],
    "SharedSize": -1,
    "Size": 3262892,
    "VirtualSize": 8138752
  },
  {
    "Containers": -1,
    "Created": 1675087699,
    "Id": "sha256:f271e74b17ced29b915d351685fd4644785c6d1559dd1f2d4189a5e851ef753a",
    "Labels": null,
    "ParentId": "",
    "RepoDigests": [
      "<none>@<none>"
    ],
    "RepoTags": [
      "🐳😄🐱"
    ],
    "SharedSize": -1,
    "Size": 3262892,
    "VirtualSize": 8138752
  }
]
```

But at least that looks somewhat better than overriding it (as the API still returns that information).
Possibly we could add some warning to the CLI that warns about a tag that's not a valid reference?